base/ca/src/main/java/com/netscape/cmscore/dbs/CertificateRepository.java

```diff
@@ -585,62 +585,29 @@ public BigInteger getLastSerialNumberInRange(BigInteger serial_low_bound, BigInt
         }
         logger.debug("CertificateRepository: getLastSerialNumberInRange  mEnableRandomSerialNumbers="+mEnableRandomSerialNumbers);
 
-        String ldapfilter = "("+CertRecord.ATTR_CERT_STATUS+"=*"+")";
-
-        String[] attrs = null;
-
-        CertRecordList recList = findCertRecordsInList(ldapfilter, attrs, serial_upper_bound.toString(10), "serialno", 5 * -1);
+        String ldapfilter = "(&("+CertRecord.ATTR_CERT_STATUS+"=*"+")("+CertRecord.ATTR_ID+"<="+serial_upper_bound+"))";
 
-        int size = recList.getSize();
+        String[] attrs = {CertRecord.ATTR_ID, "objectclass"};
 
-        logger.debug("CertificateRepository:getLastSerialNumberInRange: recList size " + size);
-
-        if (size <= 0) {
-            logger.debug("CertificateRepository:getLastSerialNumberInRange: index may be empty");
-
-            BigInteger ret = new BigInteger(serial_low_bound.toString(10));
-
-            ret = ret.subtract(BigInteger.ONE);
-            logger.debug("CertificateRepository:getLastCertRecordSerialNo: returning " + ret);
-            return ret;
-        }
-        int ltSize = recList.getSizeBeforeJumpTo();
+        RecordPagedList<CertRecord> certRecords = findPagedCertRecords(ldapfilter, attrs, "-serialno");
+        Iterator<CertRecord> iRecs = certRecords.iterator();
 
-        logger.debug("CertificateRepository:getLastSerialNumberInRange: ltSize " + ltSize);
 
-        CertRecord curRec = null;
-
-        int i;
-        Object obj = null;
-
-        for (i = 0; i < 5; i++) {
-            obj = recList.getCertRecord(i);
-
-            if (obj != null) {
-                curRec = (CertRecord) obj;
-
-                BigInteger serial = curRec.getSerialNumber();
-
-                logger.debug("CertificateRepository:getLastCertRecordSerialNo:  serialno  " + serial);
-
-                if (((serial.compareTo(serial_low_bound) == 0) || (serial.compareTo(serial_low_bound) == 1)) &&
-                        ((serial.compareTo(serial_upper_bound) == 0) || (serial.compareTo(serial_upper_bound) == -1))) {
-                    logger.debug("getLastSerialNumberInRange returning: " + serial);
-                    if (modeChange && mEnableRandomSerialNumbers) {
-                        mCounter = serial.subtract(serial_low_bound).add(BigInteger.ONE);
-                        logger.debug("getLastSerialNumberInRange mCounter: " + mCounter);
-                    }
-                    return serial;
+        if (iRecs.hasNext()) {
+            CertRecord firstRec = iRecs.next();
+            BigInteger serial = firstRec.getSerialNumber();
+            if (serial.compareTo(serial_low_bound) > 0 &&
+                    serial.compareTo(serial_upper_bound) <= 0) {
```

Comment:
I think the lower bound is inclusive too, so it should be `serial.compareTo(serial_low_bound) >= 0`.